sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetSchemasOperation.scala

```diff
@@ -77,7 +77,8 @@ private[hive] class SparkGetSchemasOperation(
 
       val globalTempViewDb = sqlContext.sessionState.catalog.globalTempViewManager.database
       val databasePattern = Pattern.compile(CLIServiceUtils.patternToRegex(schemaName))
-      if (databasePattern.matcher(globalTempViewDb).matches()) {
+      if (schemaName == null || schemaName.isEmpty ||
```

Comment:
Won't this break upstream client applications that depend on this behavior?

Reply:
https://docs.oracle.com/javase/7/docs/api/java/sql/DatabaseMetaData.html#getSchemas(java.lang.String,%20java.lang.String)
`schemaPattern - a schema name; must match the schema name as it is stored in the database; null means schema name should not be used to narrow down the search.`
This doc doesn't mention empty string, but if it's treated as a pattern, it should default to empty string not matching anything.
schemaName == null is already handled to match everything in patternToRegex:
```
  public static String patternToRegex(String pattern) {
    if (pattern == null) {
      return ".*";
    } else {
```

So the current behaviour seems to be consistent with JDBC documentation?

Reply:
https://docs.oracle.com/javase/7/docs/api/java/sql/DatabaseMetaData.html#getTables(java.lang.String,%20java.lang.String,%20java.lang.String,%20java.lang.String[])
`schemaPattern - a schema name pattern; must match the schema name as it is stored in the database; "" retrieves those without a schema; null means that the schema name should not be used to narrow the search`
The behaviour for getTables treats "" as no schema (e.g. local temp views), not all schemas, so it seems consistent that getSchemas wouldn't treat "" as "all schemas".

Reply:
Thanks @juliuszsompolski 

The java doc makes sense to me.

FYI, the current behavior in hive treats "" as wildcard,

```java
  /**
   * Convert wildchars and escape sequence of schema pattern from JDBC format to datanucleous/regex
   * The schema pattern treats empty string also as wildchar
   */

```
https://github.com/apache/hive/blob/14e3f19d027858a6f36ea6d41a3e8e45625f91bf/service/src/java/org/apache/hive/service/cli/operation/MetadataOperation.java#L75

We followed the hive behavior so far after we inlined the hive codes and before we made our own meta operations.

But now, "" is treated as wildcard for all schemas except ”global_temp“.

I guess we should either make it not a wildcard at all as JDBC API doc says or make it as same as null just like hive

Reply:
> we should either make it not a wildcard at all as JDBC API doc says'

Is it really possible? We call the hive API `listDatabases` to do it, and we can't change the hive metastore behavior.

Reply:
@cloud-fan 
this behaviour is changed by convertSchemaPattern in https://github.com/apache/hive/blob/14e3f19d027858a6f36ea6d41a3e8e45625f91bf/service/src/java/org/apache/hive/service/cli/operation/MetadataOperation.java#L75
```
    if ((pattern == null) || pattern.isEmpty()) {
      return convertPattern("%", true);
    } else {
      return convertPattern(pattern, true);
    }
```
it explicitly turns "" into a "%" pattern.
Then we use that to call listDatabases:
```
      val schemaPattern = convertSchemaPattern(schemaName)
      sqlContext.sessionState.catalog.listDatabases(schemaPattern).foreach { dbName =>
        rowSet.addRow(Array[AnyRef](dbName, DEFAULT_HIVE_CATALOG))
      }
```
but then for globalTempViews, we are not using the converted schemaPattern, but schemaName, which will not treat "" as wildcard.
```
      val globalTempViewDb = sqlContext.sessionState.catalog.globalTempViewManager.database
      val databasePattern = Pattern.compile(CLIServiceUtils.patternToRegex(schemaName))
      if (databasePattern.matcher(globalTempViewDb).matches()) {
```

described by @yaooqinn like that, it does seem to be inconsistent behaviour and just a omission bug: we should treat the pattern for global temp views the same way... JDBC standard would say that it should not be a wildcard, but Hive seems to be treating it as wildcard.
Treating it as empty would be consistent with JDBC standard. Treating it as a wildcard would be a smaller behaviour change, and consistent with Hive.
@bogdanghit @wangyum @AngersZhuuuu what do you think?

Reply:
Given that changing from Hive behavior to JDBC standard is a very breaking change ('' is not wildcard anymore), I'd prefer to fix this small behavior for global temp view here.

Reply:
sounds good. thanks a lot for the context added here @juliuszsompolski.